File: web/src/engine/phenoage.ts
/**
 * PhenoAge (Levine 2018) — a TypeScript port of Lane A's `matlab/engine/phenoage.m`.
 *
 * Every constant comes from `web/public/engine/phenoage.json` (A's export). Nothing here
 * re-derives a coefficient; the file is read and evaluated. The port must reproduce
 * `matlab/tests/vectors.json` to within 0.05 years (see `phenoage.test.ts`).
 *
 * Model (phenoage_constants.m):
 *   xb       = intercept + sum(coef_i * term_i) + coef_age * age,  term_crp = ln(max(crp, floor))
 *   M        = 1 - exp(-exp(xb) * (exp(t_months * gamma) - 1) / gamma)     (10-year mortality)
 *   PhenoAge = offset + ln(-a * ln(1 - M)) / k  ==  A + xb / k               (affine in xb)
 *
 * The affine form gives an exact per-analyte waterfall:
 *   age + cohort_offset + sum(analyte years) == phenoage
 * where cohort_offset = PhenoAge(reference person, age) - age at the user's exact age and
 * each analyte's years are coef * (term(x) - term(reference)) / k.
 */

export type Sex = 'M' | 'F';

export const ANALYTES = [
  'albumin',
  'creatinine',
  'glucose',
  'crp',
  'lymph_pct',
  'mcv',
  'rdw',
  'alp',
  'wbc',
] as const;
export type AnalyteKey = (typeof ANALYTES)[number];

type AnalyteRecord = Record<AnalyteKey, number>;
type BandTable = Record<Sex, Record<string, AnalyteRecord>>;

/** The shape of `phenoage.json` (contracts.md section 1 plus A's v8 additive keys). */
export interface PhenoAgeData {
  version: number;
  source?: string;
  units: Record<AnalyteKey | 'age', string>;
  coefficients: Record<Exclude<AnalyteKey, 'crp'> | 'ln_crp' | 'age', number>;
  intercept: number;
  gamma: number;
  t_months: number;
  k: number;
  a?: number;
  offset?: number;
  affine?: { A: number; rule?: string };
  crp_floor_mgdL: number;
  crp_acute_mgdL: number;
  reference_by_age_sex: BandTable;
  imputation_sd_by_age_sex: BandTable;
  cohort_offset_by_age_sex?: Record<Sex, Record<string, number>>;
  cv: Record<AnalyteKey, { within: number; analytical: number }>;
  critical_ranges: { glucose_mgdL: [number, number]; wbc: [number, number]; creatinine_x_ref_high: number };
  creatinine_ref_high_default_umolL?: number;
  labels?: { estimate?: string; critical?: string; imputed?: string };
}

/** Lab values in `phenoage.json` units (the API's `si_value`). Missing, null, or NaN entries are imputed. */
export type PhenoAgeValues = Partial<Record<AnalyteKey, number | null | undefined>>;

export interface PhenoAgeOptions {
  /** true = fasting. false or null/undefined (unknown) treats glucose as non-fasting: imputed, "8 of 9 markers". */
  fasting?: boolean | null;
  /** The report's upper reference limit for creatinine (umol/L), for the "2 x ref high" critical rule. */
  creatinineRefHigh?: number | null;
}

export interface PhenoAgeFlags {
  critical: boolean;
  criticalAnalytes: AnalyteKey[];
  crpAcute: boolean;
  nonFasting: boolean;
}

export type Waterfall = { cohort_offset: number } & AnalyteRecord;

export interface PhenoAgeResult {
  /** Years; null when a critical flag suppresses the number ("see a clinician first"). */
  phenoage: number | null;
  /** The non-affine route (mortality score -> age), for self-checks. Equals `phenoage` to 1e-9. */
  phenoageDirect: number | null;
  /** +/- years, 1 SD. Imputed analytes contribute the population SD, so the band widens. */
  band: number;
  waterfall: Waterfall;
  /** The nine inputs actually used, after imputation. */
  inputs: AnalyteRecord;
  imputed: AnalyteKey[];
  markersUsed: number;
  flags: PhenoAgeFlags;
  xb: number;
  mortality10y: number;
  age: number;
  sex: Sex;
  ageBand: string;
  /** "Estimate, not diagnosis", or the critical label from phenoage.json. */
  label: string;
}

/** Glucose unit conversion for the critical-range check only (mg/dL per mmol/L). Not a model coefficient. */
const MGDL_PER_MMOL_GLUCOSE = 18.016;

const DEFAULT_LABELS = { estimate: 'Estimate, not diagnosis', critical: 'See a clinician first', imputed: '8 of 9 markers' };

/**
 * Map an age to the NHANES norm band label used as a key in `reference_by_age_sex`
 * ('20-29' ... '70-79', '80+'). Ages under 20 use the lowest band (PhenoAge is an adult clock).
 */
export function ageBand(age: number, data: PhenoAgeData): string {
  const labels = Object.keys(data.reference_by_age_sex.M ?? data.reference_by_age_sex.F ?? {});
  const bands = labels
    .map((label) => {
      const open = label.endsWith('+');
      const [lo, hi] = open ? [Number(label.slice(0, -1)), Infinity] : label.split('-').map(Number);
      return { label, lo, hi: open ? hi : hi + 1 }; // '30-39' covers [30, 40)
    })
    .filter((b) => Number.isFinite(b.lo))
    .sort((a, b) => a.lo - b.lo);
  if (bands.length === 0) throw new Error('phenoage.json has no age bands');
  const a = Math.max(age, bands[0].lo);
  const hit = bands.find((b) => a >= b.lo && a < b.hi) ?? bands[bands.length - 1];
  return hit.label;
}

/** PhenoAge = A + xb / k. Prefer A's exported `affine.A`; otherwise derive it from the same constants. */
function affineA(data: PhenoAgeData): number {
  if (data.affine?.A !== undefined) return data.affine.A;
  if (data.a === undefined || data.offset === undefined) {
    throw new Error('phenoage.json needs affine.A or (a, offset)');
  }
  const logTerm = Math.log(data.a) + Math.log((Math.exp(data.t_months * data.gamma) - 1) / data.gamma);
  return data.offset + logTerm / data.k;
}

function coefficient(a: AnalyteKey, data: PhenoAgeData): number {
  return a === 'crp' ? data.coefficients.ln_crp : data.coefficients[a];
}

/** One term coef_i * f(x_i); CRP is floored, then logged. */
function modelTerm(a: AnalyteKey, value: number, data: PhenoAgeData): number {
  if (a === 'crp') return data.coefficients.ln_crp * Math.log(Math.max(value, data.crp_floor_mgdL));
  return data.coefficients[a] * value;
}

function linearPredictor(x: AnalyteRecord, age: number, data: PhenoAgeData): number {
  let xb = data.intercept + data.coefficients.age * age;
  for (const a of ANALYTES) xb += modelTerm(a, x[a], data);
  return xb;
}

function isMeasured(v: number | null | undefined): v is number {
  return typeof v === 'number' && Number.isFinite(v);
}

export function computePhenoAge(
  values: PhenoAgeValues,
  age: number,
  sex: Sex,
  data: PhenoAgeData,
  opts: PhenoAgeOptions = {},
): PhenoAgeResult {
  if (sex !== 'M' && sex !== 'F') throw new Error("sex must be 'M' or 'F'");
  if (!Number.isFinite(age)) throw new Error('age must be a number');
  const fasting = opts.fasting === true;
  const band = ageBand(age, data);
  const ref = data.reference_by_age_sex[sex]?.[band];
  const sdp = data.imputation_sd_by_age_sex[sex]?.[band];
  if (!ref || !sdp) throw new Error(`phenoage.json has no norms for ${sex} ${band}`);
  const labels = { ...DEFAULT_LABELS, ...(data.labels ?? {}) };

  // 1. Assemble the nine inputs, imputing where needed.
  const x = {} as AnalyteRecord;
  const imputed: AnalyteKey[] = [];
  const flags: PhenoAgeFlags = { critical: false, criticalAnalytes: [], crpAcute: false, nonFasting: false };
  for (const a of ANALYTES) {
    const v = values[a];
    let have = isMeasured(v);
    if (a === 'glucose' && have && !fasting) {
      have = false; // non-fasting (or unknown) glucose is imputed, greyed, "8 of 9 markers"
      flags.nonFasting = true;
    }
    if (have) {
      x[a] = v as number;
    } else {
      x[a] = ref[a];
      imputed.push(a);
    }
  }
  const measured = (a: AnalyteKey) => !imputed.includes(a);
  flags.crpAcute = measured('crp') && x.crp > data.crp_acute_mgdL;

  // 2. Critical ranges: suppress the number, "see a clinician first".
  const crit: AnalyteKey[] = [];
  const glucoseMgdl = x.glucose * MGDL_PER_MMOL_GLUCOSE;
  const [gLo, gHi] = data.critical_ranges.glucose_mgdL;
  if (measured('glucose') && (glucoseMgdl < gLo || glucoseMgdl > gHi)) crit.push('glucose');
  const [wLo, wHi] = data.critical_ranges.wbc;
  if (measured('wbc') && (x.wbc < wLo || x.wbc > wHi)) crit.push('wbc');
  const creatinineRefHigh =
    isMeasured(opts.creatinineRefHigh) && opts.creatinineRefHigh > 0
      ? opts.creatinineRefHigh
      : (data.creatinine_ref_high_default_umolL ?? 110);
  if (measured('creatinine') && x.creatinine > data.critical_ranges.creatinine_x_ref_high * creatinineRefHigh) {
    crit.push('creatinine');
  }
  flags.critical = crit.length > 0;
  flags.criticalAnalytes = crit;

  // 3. Linear predictor, both routes.
  const A = affineA(data);
  const xb = linearPredictor(x, age, data);
  const paAffine = A + xb / data.k;
  const mortality = 1 - Math.exp((-Math.exp(xb) * (Math.exp(data.t_months * data.gamma) - 1)) / data.gamma);
  const paDirect =
    data.a !== undefined && data.offset !== undefined
      ? data.offset + Math.log(-data.a * Math.log(1 - mortality)) / data.k
      : paAffine;

  // 4. Waterfall: years relative to the age-sex reference person, at the user's exact age.
  const xbRef = linearPredictor(ref, age, data);
  const waterfall = { cohort_offset: A + xbRef / data.k - age } as Waterfall;
  for (const a of ANALYTES) waterfall[a] = (modelTerm(a, x[a], data) - modelTerm(a, ref[a], data)) / data.k;

  // 5. Band: 1 SD from CV (measured) or the population SD of the term (imputed).
  let variance = 0;
  for (const a of ANALYTES) {
    const dydx = coefficient(a, data) / data.k; // years per unit of the model term
    let sdTerm: number;
    if (measured(a)) {
      const cv = data.cv[a];
      const cvTotal = Math.hypot(cv.within, cv.analytical);
      sdTerm = a === 'crp' ? cvTotal : cvTotal * x[a]; // d(ln x) = dx / x
    } else {
      sdTerm = sdp[a];
    }
    variance += (dydx * sdTerm) ** 2;
  }

  // 6. Output.
  return {
    phenoage: flags.critical ? null : paAffine,
    phenoageDirect: flags.critical ? null : paDirect,
    band: Math.sqrt(variance),
    waterfall,
    inputs: x,
    imputed,
    markersUsed: ANALYTES.length - imputed.length,
    flags,
    xb,
    mortality10y: mortality,
    age,
    sex,
    ageBand: band,
    label: flags.critical ? labels.critical : labels.estimate,
  };
}

/** age + cohort_offset + sum(analyte years); equals `phenoage` to floating point when not suppressed. */
export function waterfallSum(age: number, waterfall: Waterfall): number {
  let s = age + waterfall.cohort_offset;
  for (const a of ANALYTES) s += waterfall[a];
  return s;
}

File: matlab/engine/phenoage.m
function out = phenoage(values, age, sex, opts)
%PHENOAGE  Levine 2018 phenotypic age with an exact per-analyte waterfall and a +/- band.
%
%   out = phenoage(values, age, sex)
%   out = phenoage(values, age, sex, opts)
%
%   values : struct with any of albumin (g/L), creatinine (umol/L), glucose
%            (mmol/L), crp (mg/dL), lymph_pct (%), mcv (fL), rdw (%), alp (U/L),
%            wbc (10^9/L). Missing fields or NaN are imputed from the age-sex
%            NHANES median and listed in out.imputed ("8 of 9 markers").
%   age    : chronological age in years.        sex : 'M' or 'F'.
%   opts.fasting              true / false / [] (unknown). Anything but true
%                             treats glucose as non-fasting: greyed and imputed.
%   opts.norms                norms struct (default: load_norms()).
%   opts.creatinine_ref_high  the report's upper reference limit (umol/L) for
%                             the critical rule "creatinine > 2 x ref high".
%
%   out.phenoage          years (NaN when a critical flag suppresses it)
%   out.band              +/- years, 1 SD propagated from within-subject and
%                         analytical CV; imputed analytes contribute the
%                         population SD instead, so the band widens.
%   out.waterfall         cohort_offset plus one field per analyte, in years:
%                         age + cohort_offset + sum(analyte years) == phenoage
%                         exactly (affine identity, see phenoage_constants.m).
%   out.imputed           cellstr; out.markers_used = 9 - numel(out.imputed)
%   out.flags             critical, critical_analytes, crp_acute, non_fasting
%   out.xb, out.mortality_10y, out.phenoage_direct   the non-affine route, for self-checks

if nargin < 4, opts = struct(); end
C = phenoage_constants();
if ~isfield(opts, 'norms') || isempty(opts.norms), opts.norms = load_norms(); end
if ~isfield(opts, 'fasting'), opts.fasting = true; end
fasting = islogical(opts.fasting) && isscalar(opts.fasting) && opts.fasting;
sex = upper(char(sex));
assert(any(strcmp(sex, {'M','F'})), 'sex must be ''M'' or ''F''');
[bandLabel, bandKey] = age_band(age);
ref = opts.norms.reference.(sex).(bandKey);      % age-sex medians, canonical units
sdp = opts.norms.imputation_sd.(sex).(bandKey);  % SD of each model term in the band

% --- 1. Assemble the nine inputs, imputing where needed --------------------
x = struct(); imputed = {};
flags = struct('critical', false, 'critical_analytes', {{}}, 'crp_acute', false, 'non_fasting', false);
for i = 1:numel(C.analytes)
    a = C.analytes{i};
    have = isfield(values, a) && ~isempty(values.(a)) && ~isnan(values.(a));
    if strcmp(a, 'glucose') && have && ~fasting
        have = false; flags.non_fasting = true;     % rule 2: non-fasting glucose is imputed
    end
    if have
        x.(a) = double(values.(a));
    else
        x.(a) = ref.(a); imputed{end+1} = a; %#ok<AGROW>
    end
end
measured = @(a) ~any(strcmp(imputed, a));
flags.crp_acute = measured('crp') && x.crp > C.crp_acute_mgdL;

% --- 2. Critical ranges (rule 4: suppress the number, "see a clinician first")
crit = {};
g_mgdL = x.glucose * C.mgdL_per_mmolL_glucose;
if measured('glucose') && (g_mgdL < C.critical_ranges.glucose_mgdL(1) || g_mgdL > C.critical_ranges.glucose_mgdL(2))
    crit{end+1} = 'glucose';
end
if measured('wbc') && (x.wbc < C.critical_ranges.wbc(1) || x.wbc > C.critical_ranges.wbc(2))
    crit{end+1} = 'wbc';
end
crh = C.creatinine_ref_high_default_umolL;
if isfield(opts, 'creatinine_ref_high') && ~isempty(opts.creatinine_ref_high), crh = opts.creatinine_ref_high; end
if measured('creatinine') && x.creatinine > C.critical_ranges.creatinine_x_ref_high * crh
    crit{end+1} = 'creatinine';
end
flags.critical = ~isempty(crit);
flags.critical_analytes = crit;

% --- 3. Linear predictor, both routes ---------------------------------------
xb = linear_predictor(x, age, C);
pa_affine = C.A + xb / C.k;
M = 1 - exp(-exp(xb) * (exp(C.t_months * C.gamma) - 1) / C.gamma);
pa_direct = C.offset + log(-C.a * log(1 - M)) / C.k;

% --- 4. Waterfall: years relative to the age-sex reference person -----------
xb_ref = linear_predictor(ref, age, C);
wf = struct('cohort_offset', (C.A + xb_ref / C.k) - age);   % PhenoAge(reference, age) - age
for i = 1:numel(C.analytes)
    a = C.analytes{i};
    wf.(a) = (model_term(a, x.(a), C) - model_term(a, ref.(a), C)) / C.k;
end
s = age + wf.cohort_offset;
for i = 1:numel(C.analytes), s = s + wf.(C.analytes{i}); end
assert(abs(s - pa_affine) < 1e-9, 'waterfall does not reconcile');

% --- 5. Band: 1 SD from CV (measured) or population SD of the term (imputed)
v = 0;
for i = 1:numel(C.analytes)
    a = C.analytes{i};
    dydx = C.coefficients.(coef_name(a)) / C.k;     % years per unit of the model term
    if measured(a)
        cvt = hypot(C.cv.(a).within, C.cv.(a).analytical);
        if strcmp(a, 'crp'), sd_term = cvt; else, sd_term = cvt * x.(a); end   % d(ln x) = dx/x
    else
        sd_term = sdp.(a);
    end
    v = v + (dydx * sd_term)^2;
end

% --- 6. Output ---------------------------------------------------------------
out = struct();
out.phenoage        = pa_affine;
out.phenoage_direct = pa_direct;
out.band            = sqrt(v);
out.waterfall       = wf;
out.inputs          = x;
out.imputed         = imputed;
out.markers_used    = numel(C.analytes) - numel(imputed);
out.flags           = flags;
out.xb              = xb;
out.mortality_10y   = M;
out.age             = age;
out.sex             = sex;
out.age_band        = bandLabel;
out.label           = 'Estimate, not diagnosis';
if flags.critical
    out.phenoage = NaN; out.phenoage_direct = NaN;
    out.label = 'See a clinician first';
end
end

% ---------------------------------------------------------------------------
function xb = linear_predictor(x, age, C)
xb = C.intercept + C.coefficients.age * age;
for i = 1:numel(C.analytes)
    a = C.analytes{i};
    xb = xb + model_term(a, x.(a), C);
end
end

function t = model_term(a, val, C)
% One term coef_i * f(x_i); CRP is floored, then logged.
if strcmp(a, 'crp')
    t = C.coefficients.ln_crp * log(max(val, C.crp_floor_mgdL));
else
    t = C.coefficients.(a) * val;
end
end

function cf = coef_name(a)
if strcmp(a, 'crp'), cf = 'ln_crp'; else, cf = a; end
end
